File: output/ChatGPT/avatar/Python/Java/atcoder_ABC133_E.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

public class atcoder_ABC133_E {
    static long mod = (long) (1e9 + 7);
    static Map<Integer, List<Integer>> graph;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int K = scanner.nextInt();
        graph = new HashMap<>();

        for (int i = 0; i < N - 1; i++) {
            int a = scanner.nextInt();
            int b = scanner.nextInt();
            if (!graph.containsKey(a - 1)) {
                graph.put(a - 1, new ArrayList<>());
            }
            if (!graph.containsKey(b - 1)) {
                graph.put(b - 1, new ArrayList<>());
            }
            graph.get(a - 1).add(b - 1);
            graph.get(b - 1).add(a - 1);
        }

        long ans = dfs(0, -1, K);
        System.out.println(ans);
    }

    static long dfs(int fro, int to, int f) {
        long caseNum = f;
        int colorNum = (f == K) ? K - 1 : K - 2;
        for (int u : graph.get(fro)) {
            if (u == to) {
                continue;
            }
            long pat = dfs(u, fro, colorNum);
            if (pat == 0) {
                caseNum = 0;
                break;
            }
            caseNum = (caseNum * pat) % mod;
            colorNum -= 1;
        }
        return caseNum;
    }
}

File: output/ChatGPT/avatar/Python/C++/atcoder_ABC133_E.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> graph;
int N, K;
const int mod = 1e9 + 7;

int dfs(int fro, int to, int f) {
    int caseNum = f;
    int colorNum = (f == K) ? K - 1 : K - 2;
    for(int u : graph[fro]) {
        if(u == to) {
            continue;
        }
        int pat = dfs(u, fro, colorNum);
        if(pat == 0) {
            caseNum = 0;
            break;
        }
        caseNum = (1LL * caseNum * pat) % mod;
        colorNum -= 1;
    }
    return caseNum;
}

int main() {
    cin >> N >> K;
    graph.resize(N);
    for(int i = 0; i < N - 1; i++) {
        int a, b;
        cin >> a >> b;
        graph[a - 1].push_back(b - 1);
        graph[b - 1].push_back(a - 1);
    }
    int ans = dfs(0, -1, K);
    cout << ans << endl;

    return 0;
}